File: model_upgrade/branch_bound.py
import time

from model_upgrade.bitsets import neighbor_sets_to_masks

# Check the wall clock every this many expand() calls (perf_counter is not free).
_TIME_CHECK_INTERVAL = 2048


def branch_and_bound_max_clique(
    neighbor_sets: list[set[int]],
    lower_bound: list[int],
    deadline: float,
) -> list[int]:
    """
    Time-limited bitset branch-and-bound (Tomita MCS-style).

    Operates on integer bitmasks with a greedy-coloring upper bound, branching
    vertices in non-increasing color order so the first prune cuts the tail.
    """
    n = len(neighbor_sets)
    if n == 0:
        return []

    masks = neighbor_sets_to_masks(neighbor_sets)
    best = list(lower_bound)
    best_size = len(best)

    candidate_mask = 0
    for v in range(n):
        if len(neighbor_sets[v]) >= best_size - 1:
            candidate_mask |= 1 << v

    counter = [0]
    timed_out = [False]

    def expand(r_list: list[int], r_size: int, candidates: int) -> None:
        if timed_out[0]:
            return

        counter[0] += 1
        if counter[0] >= _TIME_CHECK_INTERVAL:
            counter[0] = 0
            if time.perf_counter() >= deadline:
                timed_out[0] = True
                return

        # Greedy coloring of `candidates`; produce vertices in color order.
        order: list[int] = []
        color_of: list[int] = []
        uncolored = candidates
        color = 0
        while uncolored:
            color += 1
            available = uncolored
            while available:
                low = available & -available
                vertex = low.bit_length() - 1
                order.append(vertex)
                color_of.append(color)
                uncolored &= ~low
                available &= ~low
                available &= ~masks[vertex]

        nonlocal best, best_size
        for i in range(len(order) - 1, -1, -1):
            if timed_out[0]:
                return
            if r_size + color_of[i] <= best_size:
                return

            vertex = order[i]
            new_size = r_size + 1
            if new_size > best_size:
                best = r_list + [vertex]
                best_size = new_size

            next_candidates = candidates & masks[vertex]
            if next_candidates:
                expand(r_list + [vertex], new_size, next_candidates)

            candidates &= ~(1 << vertex)

    expand([], 0, candidate_mask)
    return best

File: crates/model_upgrade_core/src/branch_bound.rs
use std::collections::HashSet;
use std::time::Instant;

use crate::bitsets::{neighbor_sets_to_masks, BitMask};

const TIME_CHECK_INTERVAL: u64 = 2048;

pub fn branch_and_bound_max_clique(
    neighbor_sets: &[HashSet<usize>],
    lower_bound: &[usize],
    deadline: Instant,
) -> Vec<usize> {
    let n = neighbor_sets.len();
    if n == 0 {
        return vec![];
    }

    let masks = neighbor_sets_to_masks(
        &neighbor_sets
            .iter()
            .map(|s| {
                let mut v: Vec<usize> = s.iter().copied().collect();
                v.sort_unstable();
                v
            })
            .collect::<Vec<_>>(),
    );

    let mut best = lower_bound.to_vec();
    let mut best_size = best.len();

    let mut candidate_mask = BitMask::empty();
    for v in 0..n {
        if neighbor_sets[v].len() >= best_size.saturating_sub(1) {
            candidate_mask.set(v);
        }
    }

    struct SearchState<'a> {
        masks: &'a [BitMask],
        deadline: Instant,
        best: &'a mut Vec<usize>,
        best_size: &'a mut usize,
        counter: u64,
        timed_out: bool,
    }

    fn expand(state: &mut SearchState<'_>, r_list: &mut Vec<usize>, r_size: usize, mut candidates: BitMask) {
        if state.timed_out {
            return;
        }

        state.counter += 1;
        if state.counter >= TIME_CHECK_INTERVAL {
            state.counter = 0;
            if Instant::now() >= state.deadline {
                state.timed_out = true;
                return;
            }
        }

        let mut order: Vec<usize> = Vec::new();
        let mut color_of: Vec<usize> = Vec::new();
        let mut uncolored = candidates;
        let mut color = 0usize;

        while !uncolored.is_empty() {
            color += 1;
            let mut available = uncolored;
            while !available.is_empty() {
                let vertex = available.lowest_bit().unwrap();
                order.push(vertex);
                color_of.push(color);
                uncolored.clear(vertex);
                available.clear(vertex);
                available.and_not_assign(&state.masks[vertex]);
            }
        }

        for i in (0..order.len()).rev() {
            if state.timed_out {
                return;
            }
            if r_size + color_of[i] <= *state.best_size {
                return;
            }

            let vertex = order[i];
            let new_size = r_size + 1;
            if new_size > *state.best_size {
                *state.best = r_list.iter().copied().chain(std::iter::once(vertex)).collect();
                *state.best_size = new_size;
            }

            let next_candidates = candidates.and(state.masks[vertex]);
            if !next_candidates.is_empty() {
                r_list.push(vertex);
                expand(state, r_list, new_size, next_candidates);
                r_list.pop();
            }

            candidates.clear(vertex);
        }
    }

    let mut r_list = Vec::new();
    let mut state = SearchState {
        masks: &masks,
        deadline,
        best: &mut best,
        best_size: &mut best_size,
        counter: 0,
        timed_out: false,
    };
    expand(&mut state, &mut r_list, 0, candidate_mask);
    best
}

/// Maximum independent set via Tomita-style branch-and-bound with greedy coloring bound.
pub fn branch_and_bound_max_independent_set(
    neighbor_sets: &[HashSet<usize>],
    lower_bound: &[usize],
    deadline: Instant,
) -> Vec<usize> {
    let n = neighbor_sets.len();
    if n == 0 {
        return vec![];
    }

    let masks = neighbor_sets_to_masks(
        &neighbor_sets
            .iter()
            .map(|s| {
                let mut v: Vec<usize> = s.iter().copied().collect();
                v.sort_unstable();
                v
            })
            .collect::<Vec<_>>(),
    );

    let mut best = lower_bound.to_vec();
    let mut best_size = best.len();

    let mut candidate_mask = BitMask::all(n);
    for v in 0..n {
        if neighbor_sets[v].len() > n.saturating_sub(best_size) {
            candidate_mask.clear(v);
        }
    }

    struct SearchState<'a> {
        masks: &'a [BitMask],
        deadline: Instant,
        best: &'a mut Vec<usize>,
        best_size: &'a mut usize,
        counter: u64,
        timed_out: bool,
    }

    fn expand(state: &mut SearchState<'_>, r_list: &mut Vec<usize>, r_size: usize, mut candidates: BitMask) {
        if state.timed_out {
            return;
        }

        state.counter += 1;
        if state.counter >= TIME_CHECK_INTERVAL {
            state.counter = 0;
            if Instant::now() >= state.deadline {
                state.timed_out = true;
                return;
            }
        }

        let mut order: Vec<usize> = Vec::new();
        let mut color_of: Vec<usize> = Vec::new();
        let mut uncolored = candidates;
        let mut color = 0usize;

        while !uncolored.is_empty() {
            color += 1;
            let mut available = uncolored;
            while !available.is_empty() {
                let vertex = available.lowest_bit().unwrap();
                order.push(vertex);
                color_of.push(color);
                uncolored.clear(vertex);
                available.clear(vertex);
                available.and_not_assign(&state.masks[vertex]);
            }
        }

        for i in (0..order.len()).rev() {
            if state.timed_out {
                return;
            }
            if r_size + color_of[i] <= *state.best_size {
                return;
            }

            let vertex = order[i];
            let new_size = r_size + 1;
            if new_size > *state.best_size {
                *state.best = r_list.iter().copied().chain(std::iter::once(vertex)).collect();
                *state.best_size = new_size;
            }

            let mut next_candidates = candidates;
            next_candidates.clear(vertex);
            next_candidates.and_not_assign(&state.masks[vertex]);
            if !next_candidates.is_empty() {
                r_list.push(vertex);
                expand(state, r_list, new_size, next_candidates);
                r_list.pop();
            }

            candidates.clear(vertex);
        }
    }

    let mut r_list = Vec::new();
    let mut state = SearchState {
        masks: &masks,
        deadline,
        best: &mut best,
        best_size: &mut best_size,
        counter: 0,
        timed_out: false,
    };
    expand(&mut state, &mut r_list, 0, candidate_mask);
    best
}
